### PR TITLE
Add config option for log_level

### LINKED ISSUE #188: Add config option for the grafana-agent log level
### Enhancement Proposal

There are some issues that cause the grafana-agent itself to do some excessive logging (e.g. [this issue](https://github.com/grafana/agent/issues/3885). These are not (necessarily) bugs in the charm, but the since the charm doesn't allow me to configure the log level, I can't work around them easily.
I would like to have a config option for the charm that configures the log_level of the grafana-agent itself.

### PR DESCRIPTION
**Pull request by MichaelThamm**
Closes #188

The addition of the `log_level` charm config option allows the user to filter classes of logs based on these allowable upstream levels: `["debug", "info", "warn", "error"]`. This is useful in the event of a surge of info or debug logs that prevent the user from being able to filter for more important logs like `level=error`.

# TODO
- [x] Add tests
	- [x] Scenario test
		- [x] Scenario dir root tests deleted due to errors running on main.
		- [x] Use consistent naming of context vs. ctx
	- [x] Bump to ops[testing] instead of ops-scenario.
- [x] Set `Blocked` status and centralised status settings because of model departure but let workload continue.
	- [ ] Open an issue in Prom for the same code here.

# Testing
1. Deploy and set config
```
juju deploy zookeeper --trust
juju deploy ./grafana-agent_ubuntu-22.04-amd64.charm gagent --trust
juju relate zookeeper gagent
# Wait for active/idle
juju config gagent log_level=error
```
**Note:** If you set an invalid `log_level` it will default to `info` and `juju debug log` will show:
```
unit-gagent-5: 15:10:07.219 WARNING unit.gagent/5.juju-log Invalid loglevel: panic given, debug/info/warn/error allowed. defaulting to INFO loglevel.
```
2. Check if the Grafana Agent config file is updated
`juju ssh gagent/0 cat /etc/grafana-agent.yaml | grep log_level`
3. Check if the logs are filtered
`juju ssh gagent/0 "journalctl | grep grafana-agent"`
4. Scale the principal charm
`juju add-unit zookeeper -n 1`
5. Re-run the tests from before in the scaled scenario
- [x] All manual tests passed

### CODE REVIEW

File: src/grafana_agent.py [hunk 1]
```diff
@@ -1066,6 +1067,21 @@ def _instance_name(self) -> str:
...
+    def log_level(self) -> str:
+        """The log level configured for the charm."""
+        # Valid upstream log levels in server_config
+        # https://grafana.com/docs/agent/latest/static/configuration/server-config/#server_config
+        allowed_log_levels = ["debug", "info", "warn", "error"]
+        log_level = cast(str, self.config.get("log_level")).lower()
+
+        if log_level not in allowed_log_levels:
+            logging.warning(
+                f'Invalid loglevel: {log_level} given, {"/".join(allowed_log_levels)} allowed. defaulting to INFO loglevel.'
```
Comment (sed-i) [inline src/grafana_agent.py:1080]: Nit: A list's own `__str__` would do a decent enough job I think:

```python
>>> print("Must be one of {}".format(["info", "debug"]))
Must be one of ['info', 'debug']
```

File: src/grafana_agent.py [hunk 2]
```diff
@@ -1066,6 +1067,21 @@ def _instance_name(self) -> str:
...
+            )
+            log_level = "info"
```
Comment (sed-i) [inline src/grafana_agent.py:1082]: Can we replace the string literal with what's in e.g. ops.meta?

File: tests/scenario/test_config.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+        # THEN Juju debug-log is created with WARNING level
+        found = False
+        for log in ctx.juju_log:
+            if (
+                "WARNING" == log.level
+                and "Invalid loglevel: foo given, debug/info/warn/error allowed." in log.message
+            ):
+                found = True
+                break
```
Comment (sed-i) [inline tests/scenario/test_config.py:46]: Not sure but perhaps this would look cleaner with `any`?

**Review by sed-i — APPROVED** [review]
Good job!
A few minor comments.

File: src/grafana_agent.py (hunk 2 above)
Comment (MichaelThamm) [reply src/grafana_agent.py:1082]: [ops.meta](https://ops.readthedocs.io/en/latest/#ops.CharmMeta) does not offer config options. Is this what you meant?

**Pull request merged**